layers/water-mesh.js

```diff
@@ -200,10 +225,128 @@ export class WaterMesh extends BufferedMesh {
         this.geometryBindings.delete(key);
       }
     }
+    {
+      const physicsObject = this.physicsObjectsMap.get(key);
+
+      if (physicsObject) {
+        this.physics.removeGeometry(physicsObject);
+        this.physicsObjectsMap.delete(key);
+      }
+    }
+    {
+      const chunkKey = chunk.min.x + ',' + chunk.min.y;
+      this.chunkPhysicObjcetMap.delete(chunkKey);
+      this.waterHeightMap.delete(chunkKey);
+    }
     {
       const task = this.gpuTasks.get(key);
       task.cancel();
       this.gpuTasks.delete(key);
     }
   }
+
+  checkWaterContact(chunkPhysicObject, player, waterSurfaceHeight) {
+    // use overlapBox to check whether player contact the water
+    this.physics.enableGeometryQueries(chunkPhysicObject);
+    if (player.avatar) {
+      let collisionIds;
+      const height = player.avatar.height * 0.9;
+      const width = player.avatar.shoulderWidth
+      if (player.position.y > waterSurfaceHeight) {
+        collisionIds = this.physics.overlapBox(width, height, width, player.position, player.quaternion).objectIds;
+      }
+      else {
+        localVector.set(player.position.x, waterSurfaceHeight, player.position.z);
+        collisionIds = this.physics.overlapBox(width, height, width, localVector, player.quaternion).objectIds;
+      } 
+      for (const collisionId of collisionIds) {
+        if (collisionId === chunkPhysicObject.physicsId) {
+          this.physics.disableGeometryQueries(chunkPhysicObject);
+          return true;
+        }
+      }
+    }
+    this.physics.disableGeometryQueries(chunkPhysicObject);
+    return false;
+  }
+  getSwimDamping(player) {
+    if (this.lastSwimmingHand !== player.avatarCharacterSfx.currentSwimmingHand) {
+      this.lastSwimmingHand = player.avatarCharacterSfx.currentSwimmingHand;
+      if (player.avatarCharacterSfx.currentSwimmingHand !== null) {
+        return 1;
+      }
+    }
+    if (this.swimDamping < 4.2 && this.lastSwimmingHand) {
+      return this.swimDamping *= 1.03;
+    }
+    else {
+      return 4.2;
+    }
+  }
+  handleSwimAction(contactWater, player, waterSurfaceHeight) {
+    const swimAction = player.getAction('swim');
+    const hasSwim = !!swimAction;
+    if (contactWater) {
+      this.material.color.setHex( 0x0000ff ); // for testing
+      if (waterSurfaceHeight >= player.position.y - player.avatar.height + player.avatar.height * 0.8) { // if water is higher than player's neck

```

Comment:
Two things :
- Pls cleanup the math in the if() parentheses 
- What is 0.8 here ? use a constant variable and the name should explain what the 0.8 is

Reply:
[use constant for character head and neck](https://github.com/webaverse/heightfield/pull/6/commits/edbcf35c0ef3e869cfb821f74a2815e45ff92394)